contracts/SolidexJoint.sol

```diff
@@ -0,0 +1,370 @@
+        uint256 expectedBalanceA,
+        uint256 expectedBalanceB
+    ) external override onlyVaultManagers {
+        ISolidRouter(router).removeLiquidity(
+            tokenA,
+            tokenB,
+            stable,
+            amount,
+            0,
+            0,
+            address(this),
+            now
+        );
+        require(expectedBalanceA <= balanceOfA(), "!sandwidched");
+        require(expectedBalanceA <= balanceOfB(), "!sandwidched");
+    }
+
+    function sellCapital(
+        address _tokenFrom,
+        address _tokenTo,
+        uint256 _amountIn
+    ) internal override returns (uint256 _amountOut) {
+        uint256[] memory amounts =
+            ISolidRouter(router).swapExactTokensForTokens(
+                _amountIn,
+                0,
+                getTokenOutPathSolid(_tokenFrom, _tokenTo),
+                address(this),
+                now
+            );
+        _amountOut = amounts[amounts.length - 1];
+    }
+
+
+    function getTokenOutPathSolid(address _token_in, address _token_out)
+        internal
+        view
+        returns (ISolidRouter.route[] memory _routes)
+    {
+
+        address[] memory _path;
+        bool is_weth =
+            _token_in == address(WETH) || _token_out == address(WETH);
+        bool is_internal =
+            (_token_in == tokenA && _token_out == tokenB) ||
+                (_token_in == tokenB && _token_out == tokenA);
+        _path = new address[](is_weth || is_internal ? 2 : 3);
+        _path[0] = _token_in;
+        if (is_weth || is_internal) {
+            _path[1] = _token_out;
+        } else {
+            _path[1] = address(WETH);
+            _path[2] = _token_out;
+        }
+
+        uint256 pathLength = _path.length > 1 ? _path.length - 1 : 1;
+        _routes = new ISolidRouter.route[](pathLength);
+        for(uint i = 0; i < pathLength; i++) {
+            bool isStable = is_internal ? stable : false;
+            _routes[i] = ISolidRouter.route(_path[i], _path[i+1], isStable);
+        }
+    }
+
+    function swapReward(uint256 _rewardBal)
+        override
+        internal
+        returns (address, uint256)
+    {
+        // WARNING: NOT SELLING REWARDS! !!! 
+
+
+
+
+        if (reward == tokenA || reward == tokenB || _rewardBal == 0) {
+            return (reward, 0);
+        }
+
+        if (tokenA == WETH || tokenB == WETH) {
+            return (WETH, 0);
+        }
+
+        // Assume that position has already been liquidated
+        (uint256 ratioA, uint256 ratioB) =
+            getRatios(balanceOfA(), balanceOfB(), investedA, investedB);
+        if (ratioA >= ratioB) {
+            return (tokenB, 0);
+        }
+        return (tokenA, 0);
+    }
+
+    function getPair() internal view override returns (address) {
+        address factory = ISolidRouter(router).factory();
+        return ISolidFactory(factory).getPair(tokenA, tokenB, stable);
+    }
+
+    
+    function estimatedTotalAssetsAfterBalance()
```

Comment:
what is this used for? might as well remove if we arent swapping

Reply:
reporting estimatedTotalAssets (avoid reverts on underlying code)